### PR TITLE
Feature/332 del contac

### LINKED ISSUE #332: Deleting a 1:1 contact, does not show up in Recycle Bin, so can't be undeleted.
_Issue by **[kbromer](https://github.com/kbromer)** from Friday Oct 12, 2012 at 20:37 GMT_
_Originally opened as https://github.com/SalesforceFoundation/Contacts_and_Organizations/issues/8_

----

http://code.google.com/p/npsp/issues/detail?id=342
and
https://groups.google.com/forum/?fromgroups=#!topic/npsf/5pcBKQExlmc

### LINKED ISSUE #1678: Null Opportunity amount causes Manage Allocations to fail
https://powerofus.force.com/0D58000002EIzXX

### PR DESCRIPTION
**Pull request by njjc**
# Warning
# Info
- The Contact delete button override should be updated to use the CON_DeleteContactOverride visualforce page: Setup > Build > Customize > Contacts > Buttons, Links, and Actions > Click 'Edit' next to Delete > Select 'Visualforce Page' and choose CON_DeleteContactOverride.

# Issues
Fixes #332 
Fixes #1678

### CODE REVIEW

File: src/classes/CON_DeleteContactOverride_CTRL.cls [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        retURL = ApexPages.currentPage().getParameters().get('returl');
+        
+        //if we don't have a return URL, go back to Contacts home.
+        if (string.isBlank(retURL)) {
+            retURL = '/003/o';
+        }
+    }
+
+    /*******************************************************************************************************
+    * @description Standard controller constructor.
```
Comment (davidhabib) [inline src/classes/CON_DeleteContactOverride_CTRL.cls]: should be something like action method to handle deleting a contact and potentially its account if it is the last contact in the account.

File: src/classes/CON_DeleteContactOverride_CTRL.cls-meta.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>31.0</apiVersion>
```
Comment (davidhabib) [inline src/classes/CON_DeleteContactOverride_CTRL.cls-meta.xml]: bump to 33.

File: src/classes/CON_DeleteContactOverride_TEST.cls [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        
+        system.assertEquals(1,queryAcc.size(),'There should be one account.');
+        system.assert(queryacc[0].isDeleted,'Deleted flag should not be set.');
+
+        queryCon = [SELECT Id, isDeleted FROM Contact WHERE id = :con.id ALL ROWS];
+
+        system.assertEquals(1,queryCon.size(),'There should be one deleted contact.');
+        system.assert(queryCon[0].isDeleted,'Deleted flag should be set.');
+    }
+}
```
Comment (davidhabib) [inline src/classes/CON_DeleteContactOverride_TEST.cls]: I assume any call by any code to delete listContact or delete contact, won't go through the override.  So there is no point in testing a scenario where the account has multiple contacts, and you call delete listContact on all of them?

File: src/classes/CON_DeleteContactOverride_TEST.cls-meta.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>31.0</apiVersion>
```
Comment (davidhabib) [inline src/classes/CON_DeleteContactOverride_TEST.cls-meta.xml]: bump to 33

File: src/pages/CON_DeleteContactOverride.page-meta.xml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexPage xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>31.0</apiVersion>
```
Comment (davidhabib) [inline src/pages/CON_DeleteContactOverride.page-meta.xml]: bump to 33.

File: src/classes/CON_DeleteContactOverride_TEST.cls (hunk 3 above)
Comment (njjc) [inline src/classes/CON_DeleteContactOverride_TEST.cls]: Correct.

**Comment by njjc** [comment]
@davidhabib Made the changes you suggested, should be good to go.

**Pull request merged**